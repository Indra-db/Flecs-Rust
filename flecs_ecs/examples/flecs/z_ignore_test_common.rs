use std::sync::{Arc, Mutex};

#[cfg(feature = "flecs_nightly_tests")]
pub use capture_stdio::*;
#[cfg(feature = "flecs_nightly_tests")]
use std::io::{set_output_capture, Error};

/// Wrap `std::io::set_output_capture`
#[cfg(feature = "flecs_nightly_tests")]
pub struct OutputCapture {
    local_stream: Arc<Mutex<Vec<u8>>>,
    original: Option<Arc<Mutex<Vec<u8>>>>,
    restored: bool,
}

#[cfg(feature = "flecs_nightly_tests")]
impl Capture for OutputCapture {
    fn capture() -> Result<Self, Error> {
        let local_stream = Arc::new(Mutex::new(vec![]));
        let original = set_output_capture(Some(local_stream.clone()));
        Ok(Self {
            local_stream,
            original,
            restored: false,
        })
    }

    fn restore(&mut self) {
        assert!(!self.restored, "cannot restore it twice");

        set_output_capture(self.original.clone());
        self.restored = true;
    }
}

#[cfg(feature = "flecs_nightly_tests")]
impl OutputCapture {
    /// Get the captured output
    pub fn output(&self) -> Arc<Mutex<Vec<u8>>> {
        self.local_stream.clone()
    }

    pub fn output_string(&self) -> String {
        String::from_utf8(self.output().lock().unwrap().clone()).unwrap()
    }

    pub fn test(&self, name: String) {
        let str_output = String::from_utf8(self.output().lock().unwrap().clone()).unwrap();
        let mut settings = insta::Settings::clone_current();
        #[allow(clippy::double_parens)]
        settings._private_inner_mut().filters(
            (vec![
                (r"id: (\d+)\s", "[ID] "),
                (r"#(\d+)\s", "#[ID] "),
                (
                    r#"Group deleted: \\"([^\\"]+)\\""#,
                    "group deleted: redacted",
                ),
            ]),
        );
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix(name);
        settings.set_snapshot_path("z_ignore_test_snapshots");
        settings.bind(|| {
            insta::assert_yaml_snapshot!(str_output);
        });
    }
}

#[cfg(feature = "flecs_nightly_tests")]
impl Drop for OutputCapture {
    fn drop(&mut self) {
        if !self.restored {
            self.restore();
        }
    }
}
