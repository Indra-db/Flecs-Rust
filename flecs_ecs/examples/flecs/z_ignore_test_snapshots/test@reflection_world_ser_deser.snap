---
source: flecs_ecs/examples/flecs/z_ignore_test_common.rs
expression: str_output
---
"{\"results\":[{\"name\":\"#16\", \"id\":16, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"alloc.string.String\"}}}, {\"name\":\"#18\", \"id\":18, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.i8\"}}}, {\"name\":\"#20\", \"id\":20, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.i16\"}}}, {\"name\":\"#22\", \"id\":22, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.i32\"}}}, {\"name\":\"#24\", \"id\":24, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.i64\"}}}, {\"name\":\"#26\", \"id\":26, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.u8\"}}}, {\"name\":\"#28\", \"id\":28, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.u16\"}}}, {\"name\":\"#30\", \"id\":30, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.u32\"}}}, {\"name\":\"#32\", \"id\":32, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.u64\"}}}, {\"name\":\"#34\", \"id\":34, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.f32\"}}}, {\"name\":\"#36\", \"id\":36, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.f64\"}}}, {\"name\":\"#38\", \"id\":38, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.bool\"}}}, {\"name\":\"#40\", \"id\":40, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.char\"}}}, {\"name\":\"#42\", \"id\":42, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.uptr\"}}}, {\"name\":\"#44\", \"id\":44, \"components\":{\"flecs.core.Component\":{\"size\":16, \"alignment\":8}, \"flecs.meta.type\":{\"kind\":\"VectorType\"}, \"flecs.meta.TypeSerializer\":null, \"flecs.meta.vector\":{\"type\":\"flecs.meta.iptr\"}}}, {\"name\":\"ent_1\", \"id\":551, \"components\":{\"examples.reflection.reflection_world_ser_deser.Position\":{\"x\":10, \"y\":20}, \"examples.reflection.reflection_world_ser_deser.Velocity\":{\"x\":1, \"y\":-1}, \"(flecs.core.Identifier,flecs.core.Name)\":null}}, {\"name\":\"ent_2\", \"id\":552, \"components\":{\"examples.reflection.reflection_world_ser_deser.Position\":{\"x\":30, \"y\":40}, \"examples.reflection.reflection_world_ser_deser.Velocity\":{\"x\":-1, \"y\":1}, \"(flecs.core.Identifier,flecs.core.Name)\":null}}, {\"name\":\"Move\", \"id\":550, \"tags\":[\"flecs.core.Query\", \"flecs.system.System\", \"flecs.pipeline.OnUpdate\"],\"pairs\":{\"flecs.core.DependsOn\":\"flecs.pipeline.OnUpdate\"}, \"components\":{\"(flecs.core.Identifier,flecs.core.Name)\":null, \"(flecs.core.Poly,flecs.core.Query)\":null, \"(flecs.core.Poly,flecs.system.System)\":null}}]}\n::ent_1 moved to {x: 11, y: 19}\n::ent_2 moved to {x: 29, y: 41}\n\n::ent_1 moved to {x: 11, y: 19}\n::ent_2 moved to {x: 29, y: 41}\n"
